File: problems/problem28/Problem28.cpp
#include <iostream>

int Problem28(int n)
{
    n = (n - 1) / 2;

    int start = 3, times, sum = 3;

    for (int i = 1; i < n + 1; i++)
    {
        if (i == 1)
        {
            times = 2;
        }
        else
        {
            times = 3;
        }

        for (int j = 0; j < times + 1; j++)
        {
            start += i * 2;
            sum += start;
        }
    }
    return sum + 1; // since the loop does not account for the 1x1 matrix case, we include it here by adding the 1 (1x1 answer) here
}

int main()
{
    std::cout << Problem28(1001) << "\n";
    return 0;
}

File: problems/problem28/problem28.py
from numba import jit

# unoptimized
def Problem28(n):
    n = (n - 1) // 2
    
    # generate diagonal values:
    start = 3
    Sum = 3
    
    for i in range(1, n + 1):
        if i == 1:
            times = 2
        else:
            times = 3
            
        for _ in range(times + 1):
            start += i * 2
            Sum += start
                
    return Sum + 1

@jit
def Problem28J(n):
    n = (n - 1) // 2
    
    # generate diagonal values:
    start = 3
    Sum = 3
    
    for i in range(1, n + 1):
        if i == 1:
            times = 2
        else:
            times = 3
            
        for _ in range(times + 1):
            start += i * 2
            Sum += start
                
    return Sum + 1
  
print(Problem28(1001))
print(Problem28J(1001)) # roughly 160-223x faster
